### PR TITLE
Setup pre-commit github action and apply pre-commit to all files

### PR DESCRIPTION
**Pull request by anupambhatnagar**
## What does this PR do?
Fixes Issue [842](https://github.com/facebookresearch/fairscale/issues/842)

This PR makes the following changes:

- Moved isort, black, mypy and flake8 checks from CircleCI to Github Actions.
- Version updates in `.pre-commit-config.yaml` and `requirements-test.txt`. Made sure the version are the same in both the files.
- Updated torch version in `requirements.txt`
- Require `pre-commit >=2.15.0` in `requirements-dev.txt`
- Created a github workflow to run pre-commit. See `.github/workflows/pre-commit.yml`
- Changes due to mypy
  - `stubs/torch/nn/modules/adaptive.pyi:10` renamed namedtuple
  - `experimental/optim/dynamic_loss_scaler.py` added an else block
  - `fairscale/experimental/nn/offload.py:473` suppressed mypy error
  - removed unused "type: ignore" comments in the following files
    - `fairscale/optim/oss.py`
    - `fairscale/nn/pipe/async_schedule.py`
    - `fairscale/experimental/nn/ampnet_pipe/pipe.py`
    - `fairscale/experimental/tooling/layer_memory_tracker.py`

Although there are a lot of small changes in this PR, they are all related to formatting changes.

## Before submitting

- [x] Did you have fun?
  - Make sure you had fun coding 🙃
- [x] Did you read the [contributor guideline](https://github.com/facebookresearch/fairscale/blob/main/CONTRIBUTING.md)?
- [x] Was this discussed/approved via a Github issue? (no need for typos, doc improvements)
  - [ ] N/A
- [ ] Did you make sure to update the docs?
  - [x] N/A
- [ ] Did you write any new necessary tests?
  - [x] N/A
- [ ] Did you update the [changelog](https://github.com/facebookresearch/fairscale/blob/main/CHANGELOG.md)? (if needed)
  - [x] N/A

## PR review
Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

### CODE REVIEW

File: .github/workflows/pre-commit.yml [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+  push:
+    branches: [main]
+
+jobs:
+  pre-commit:
+    runs-on: ubuntu-latest
+    strategy:
+      matrix:
+        # make sure python versions are consistent with those used in .circleci/config.yml
+        python-version: ['3.7.12', '3.8.12']
```
Comment (anj-s) [inline .github/workflows/pre-commit.yml:14]: nit: what about python 3.9?

File: .circleci/config.yml [hunk 2]
```diff
@@ -138,7 +141,7 @@ run_mypy: &run_mypy
    - run:
        name: Run type-checking (mypy)
        command: |
-         mypy --ignore-missing-imports --scripts-are-modules --pretty .
+         mypy --no-strict-optional --ignore-missing-imports --scripts-are-modules --pretty .
```
Comment (anj-s) [inline .circleci/config.yml:144]: what does --no-strict-optional add for us?

**Comment by anj-s** [comment]
are all the formatting changes because of the pre-commit hook change?

File: .github/workflows/pre-commit.yml (hunk 1 above)
Comment (anupambhatnagar) [reply .github/workflows/pre-commit.yml:14]: good catch. there is only one job using 3.9 so I missed it.

File: .circleci/config.yml (hunk 2 above)
Comment (anupambhatnagar) [reply .circleci/config.yml:144]: From https://github.com/pre-commit/mirrors-mypy

"By default, mypy will run with mypy --ignore-missing-imports, pre-commit runs mypy from an isolated virtualenv so it won't have access to those. To change the arguments, override the args as follows:
    hooks:
    -   id: mypy
        args: [--no-strict-optional, --ignore-missing-imports]
"

**Comment by anupambhatnagar** [comment]
@anj-s yes, all formatting changes are because of pre-commit.

File: .pre-commit-config.yaml [hunk 3]
```diff
@@ -20,29 +20,31 @@ repos:
...
-    rev: 5.6.4
+    rev: 5.10.1
     hooks:
     -   id: isort
         exclude: README.md
         additional_dependencies: [toml]
 
 -   repo: https://github.com/pre-commit/mirrors-mypy
-    rev: 'v0.790'
+    rev: 'v0.910'
```
Comment (anj-s) [inline .pre-commit-config.yaml:46]: How do we ensure that we keep updating this config regularly?

**Review by anj-s — APPROVED** [review]

**Comment by anj-s** [comment]
Once the tests pass please feel free to merge. Thank you so much for the cleanup and updating the config!

nit: Do remove "WIP" in the PR title and instead add an appropriate tag.

File: .pre-commit-config.yaml (hunk 3 above)
Comment (anupambhatnagar) [reply .pre-commit-config.yaml:46]: We should have a task in our quarterly fix-it log to ensure that the CI system is working as expected. Based on my limited usage so far there are two tasks which should be done every quarter (and these count towards Better Engg as well):
1. Update the version of all the linters (black, mypy, isort, flake8 etc.)
2. Revisit tests to ensure that they are evenly split by time i.e. all three partitions take approximately the same amount of time and perhaps increase the parallelism, if needed. Here's a link which shows how to achieve this. https://circleci.com/docs/2.0/parallelism-faster-jobs/#running-split-tests. The current state can be assessed from the attached image.

circle_ci_timings

**Pull request merged**